### ISSUE TITLE
Add categories and filter function to History, Connections & Credentials

### ISSUE THREAD

**Issue by Oran-Dan** [impierce/identity-wallet#77]
### Description

As all three tabs (History, Connections & Credentials) might grow considerably during usage, it would be very convenient to be able to categorize/label and filter accordingly. 

History will have 3 categories from the start, namely 3 types of events: connected, data shared and data received.
Both credentials and connections should be able to be labeled by the user by real-life categories such as "medical", "educational", "governmental" etc. For the History, I think this metadata should also be automatically induced from the connections as they happen. So if a connection has already been categorized so should the event be.
Still the labeling option would be nice to have available in the history section too, although somewhere in the future History events should be reportable when privacy breaches occur. In this case, I assume we should be able to prove these events are untampered with and are indeed reliable proof of what happened in the past. This means these labels should live outside an immutable data-core of these events.

Furthermore, I think the labels should be the same for all three datatypes. No need to seperate them per History/Connection/Credential I think. User can always add a new label if they want to.

I might suggest prompting the user to label connections and credentials immediately as they come in. This could be done with perhaps 3 or 4 standard labels as suggested above and an option to create your own label (would also be nice if the labels have their own optional colors too) . This would prevent users from postponing it until it's already messy and a big hassle to do it.

Then there should be a label/category management tab under Settings perhaps. Where users will be able to rename/delete a label or add a different color.

Of course these specifics are up to discussion and should be agreed upon by Jelle, front-end and back-end before starting with this issue.

P.S. Should we allow for subcategories? Some people might have extensive educational/medical records with multiple universities/hospitals.

### Motivation

Adds to the user experience.

### To-do List

Back-end:
- [ ] Add optional label/category field to Credentials in the display substruct (next to is_favorite boolean).
- [ ] Add optional label/category field to Connection struct.
- [ ] Add optional label/category field to Events as metadata.
- [ ] Add Action with an enum to serve all the above three datatypes to update the labels.
- [ ] Add UserPrompt to invoke user to enter a category upon event/connecting (if this is an agreeable ux)
- [ ] Add datafield to store the different labels/categories in use. Action above should be able to automatically add a new label to this datafield when created through clicking on an instance of one of the three datatypes above.
- [ ] Add label/category management tab (under settings tab as well I might suggest) to be able to manage names/colors and add new ones.
- [ ] Create Action to handle label/category changes through the respective management tab. Changes should then of course be implemented throughout it's use in all 3 datatypes.
- [ ] Add filter buttons to History, Credential and Connection tabs and integrate this with the search & sort functionality.

Front-end:
- [ ] This to do list depends on the designs I figure so I will leave this open.

**Comment by M-Adam-Hus** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Add categories and filter function to History, Connections & Credentials](https://trello.com/c/GX5Zpumh/15-add-categories-and-filter-function-to-history-connections-credentials)
